File: pharmacy-ecommerce/scripts/fix_broken_images.mjs
/**
 * fix_broken_images.mjs
 *
 * Fase 1 — Chequeo paralelo de URLs (20 concurrent, ~5 min para 1462 productos)
 * Fase 2 — Búsqueda de reemplazo en DuckDuckGo + update en Cloud SQL
 *
 * Ejecutar desde pharmacy-ecommerce/:
 *   node scripts/fix_broken_images.mjs
 *   node scripts/fix_broken_images.mjs --check-only   (solo fase 1, guarda broken_images_log.json)
 *   node scripts/fix_broken_images.mjs --fix-only     (solo fase 2, asume broken_images_log.json ya existe)
 */

import { readFileSync, writeFileSync, existsSync } from 'fs'
import { createRequire } from 'module'
import { join } from 'path'

// ─── Setup ────────────────────────────────────────────────────────────────────
const webDir = new URL('../apps/web/', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1')
const require = createRequire(join(webDir, 'package.json'))

// Load .env.local
const envRaw = readFileSync(join(webDir, '.env.local'), 'utf-8')
const env = Object.fromEntries(
  envRaw.split('\n')
    .filter(l => l.trim() && !l.startsWith('#') && l.includes('='))
    .map(l => {
      const i = l.indexOf('=')
      const key = l.slice(0, i).trim()
      let val = l.slice(i + 1).trim()
      if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'")))
        val = val.slice(1, -1)
      return [key, val]
    })
)

// Cloud SQL
const { Connector, IpAddressTypes, AuthTypes } = require('@google-cloud/cloud-sql-connector')
const { GoogleAuth } = require('google-auth-library')
const pg = require('pg')

const PROGRESS_FILE = join(import.meta.dirname || process.cwd(), 'fix_images_progress.json')
const BROKEN_LOG    = join(import.meta.dirname || process.cwd(), 'broken_images_log.json')

const args = process.argv.slice(2)
const CHECK_ONLY = args.includes('--check-only')
const FIX_ONLY   = args.includes('--fix-only')

// ─── Helpers: parse JSON env vars stored with escape sequences in .env.local ──
function parseEnvJson(val) {
  if (!val) throw new Error('GOOGLE_SERVICE_ACCOUNT not set')
  // Direct parse (works in Vercel / production where value is stored as real JSON)
  try { return JSON.parse(val) } catch {}
  // .env.local stores it as a JSON-string-escaped value ("{\n  \"type\":...}")
  // Wrapping in quotes and double-parsing decodes \n → newline, \" → " etc.
  try { return JSON.parse(JSON.parse('"' + val + '"')) } catch (e) {
    throw new Error('Cannot parse GOOGLE_SERVICE_ACCOUNT: ' + e.message)
  }
}

// ─── DB connection ────────────────────────────────────────────────────────────
async function connectDb() {
  const credentials = parseEnvJson(env.GOOGLE_SERVICE_ACCOUNT)
  const auth = new GoogleAuth({
    credentials,
    scopes: ['https://www.googleapis.com/auth/cloud-platform'],
  })
  const connector = new Connector({ auth })
  const clientOpts = await connector.getOptions({
    instanceConnectionName: env.CLOUD_SQL_INSTANCE.trim(),
    ipType: IpAddressTypes.PUBLIC,
    authType: AuthTypes.PASSWORD,
  })
  const pool = new pg.Pool({
    ...clientOpts,
    user: env.DB_USER.trim(),
    password: env.DB_PASSWORD.trim(),
    database: env.DB_NAME.trim(),
    max: 5,
  })
  return pool
}

// ─── URL checker ──────────────────────────────────────────────────────────────
async function checkUrl(url, timeoutMs = 6000) {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutMs)
  try {
    const httpsUrl = url.startsWith('http://') ? 'https://' + url.slice(7) : url
    const res = await fetch(httpsUrl, {
      method: 'HEAD',
      signal: controller.signal,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Referer': 'https://tu-farmacia.cl/',   // simulate real browser request from our domain
        'Accept': 'image/*,*/*',
      },
      redirect: 'follow',
    })
    clearTimeout(timer)
    // 403 = hotlink blocked (broken for our domain)
    // 404 = gone
    // 5xx = server error
    return { alive: res.status < 400, status: res.status }
  } catch (e) {
    clearTimeout(timer)
    return { alive: false, status: null }  // timeout / conn error
  }
}

// ─── Parallel URL check ───────────────────────────────────────────────────────
async function checkAllUrls(products, concurrency = 20) {
  const broken = []
  const ok = []
  let done = 0
  const total = products.length
  const start = Date.now()

  // Chunk into batches
  for (let i = 0; i < total; i += concurrency) {
    const batch = products.slice(i, i + concurrency)
    const results = await Promise.all(batch.map(p => checkUrl(p.image_url)))
    for (let j = 0; j < batch.length; j++) {
      const prod = batch[j]
      const { alive, status } = results[j]
      done++
      const elapsed = ((Date.now() - start) / 1000 / 60).toFixed(1)
      const eta = done > 0 ? ((((Date.now() - start) / done) * (total - done)) / 1000 / 60).toFixed(1) : '?'
      const nameShort = (prod.name || '').slice(0, 40).padEnd(40)
      const statusStr = status ? `${status}` : 'ERR'
      process.stdout.write(`\r[${done}/${total}] (${elapsed}m ETA:${eta}m) ${nameShort} ${alive ? 'OK ' : `ROTO(${statusStr})`}   `)
      if (alive) {
        ok.push(prod)
      } else {
        broken.push({ id: prod.id, name: prod.name, laboratory: prod.laboratory, broken_url: prod.image_url, status })
      }
    }
    // Small delay between batches to avoid hammering multiple hosts at once
    await sleep(200)
  }
  console.log()
  return { broken, ok }
}

// ─── DuckDuckGo image search ──────────────────────────────────────────────────
async function searchImage(name, lab = '', retries = 2) {
  const cleanName = name.split('(')[0].trim()
  const queries = [
    `${cleanName} ${lab} farmacia`.trim(),
    `${cleanName} medicamento`,
    `${cleanName} comprimido pastilla`,
    cleanName.slice(0, 40),
  ]
  const unique = [...new Set(queries.filter(q => q.trim()))]

  for (const query of unique) {
    try {
      const encoded = encodeURIComponent(query)
      // Step 1: get VQD token
      const tokenRes = await fetch(`https://duckduckgo.com/?q=${encoded}`, {
        headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36' },
      })
      const html = await tokenRes.text()
      const match = html.match(/vqd=['"]([^'"]+)['"]/)
      if (!match) continue

      const vqd = match[1]
      await sleep(800)

      // Step 2: image results
      const imgRes = await fetch(
        `https://duckduckgo.com/i.js?l=cl-es&o=json&q=${encoded}&vqd=${vqd}&f=,,,,,&p=1`,
        { headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36' } }
      )
      if (!imgRes.ok) {
        if (imgRes.status === 429 || imgRes.status === 202) {
          if (retries > 0) {
            console.log('\n  [RATE LIMIT] Esperando 30s...')
            await sleep(30000)
            return searchImage(name, lab, retries - 1)
          }
          return null
        }
        continue
      }

      const data = await imgRes.json()
      if (data.results?.length) {
        for (const r of data.results.slice(0, 5)) {
          const img = r.image
          if (img && img.startsWith('http') && !img.includes('x-raw-image') && img.length < 500 && !img.endsWith('.gif')) {
            return img.startsWith('http://') ? 'https://' + img.slice(7) : img
          }
        }
      }
    } catch (e) {
      await sleep(1000)
    }
  }
  return null
}

// ─── Helpers ─────────────────────────────────────────────────────────────────
function sleep(ms) { return new Promise(r => setTimeout(r, ms)) }

function loadProgress() {
  try { return new Set(JSON.parse(readFileSync(PROGRESS_FILE, 'utf-8'))) } catch { return new Set() }
}
function saveProgress(ids) { writeFileSync(PROGRESS_FILE, JSON.stringify([...ids])) }
function loadBrokenLog() {
  try { return JSON.parse(readFileSync(BROKEN_LOG, 'utf-8')) } catch { return [] }
}
function saveBrokenLog(entries) { writeFileSync(BROKEN_LOG, JSON.stringify(entries, null, 2)) }

// ─── Main ─────────────────────────────────────────────────────────────────────
async function main() {
  console.log('═'.repeat(60))
  console.log('  DETECCIÓN Y ARREGLO DE IMÁGENES ROTAS — Cloud SQL')
  console.log('═'.repeat(60))
  console.log()

  const pool = await connectDb()
  console.log('✓ Conectado a Cloud SQL\n')

  let brokenList

  // ── FASE 1: Chequeo de URLs ──────────────────────────────────────────────
  if (!FIX_ONLY) {
    console.log('FASE 1: Chequeando URLs existentes...')
    const { rows: products } = await pool.query(`
      SELECT id, name, laboratory, image_url
      FROM products
      WHERE image_url IS NOT NULL AND image_url <> ''
      ORDER BY name
    `)
    console.log(`  ${products.length} productos con image_url\n`)

    const { broken, ok } = await checkAllUrls(products, 20)
    console.log(`\n  ✓ Válidas: ${ok.length} | Rotas: ${broken.length}\n`)

    saveBrokenLog(broken)
    console.log(`  → Log guardado: broken_images_log.json\n`)

    brokenList = broken
    if (CHECK_ONLY) {
      console.log('Modo --check-only: terminando aquí.')
      await pool.end()
      return
    }
  } else {
    brokenList = loadBrokenLog()
    console.log(`Modo --fix-only: ${brokenList.length} URLs rotas desde broken_images_log.json\n`)
  }

  // ── FASE 2: Búsqueda y reparación ───────────────────────────────────────
  if (brokenList.length === 0) {
    console.log('Sin imágenes rotas. ¡Todo OK!')
    await pool.end()
    return
  }

  console.log(`FASE 2: Buscando reemplazos para ${brokenList.length} imágenes rotas...`)
  console.log('  (DuckDuckGo, ~2-4s por producto)\n')

  const done = loadProgress()
  const pending = brokenList.filter(p => !done.has(p.id))
  if (done.size > 0) console.log(`  Retomando: ${done.size} ya procesados, ${pending.length} pendientes\n`)

  let fixed = 0, nulled = 0, errors = 0
  const start = Date.now()

  for (let i = 0; i < pending.length; i++) {
    const prod = pending[i]
    const elapsed = ((Date.now() - start) / 1000 / 60).toFixed(1)
    const eta = i > 0 ? ((((Date.now() - start) / i) * (pending.length - i)) / 1000 / 60).toFixed(1) : '?'
    const nameShort = prod.name.slice(0, 40).padEnd(40)
    process.stdout.write(`[${i + 1}/${pending.length}] (${elapsed}m ETA:${eta}m) ${nameShort} `)

    const newUrl = await searchImage(prod.name, prod.laboratory || '')

    if (newUrl) {
      try {
        await pool.query('UPDATE products SET image_url = $1 WHERE id = $2', [newUrl, prod.id])
        fixed++
        prod.new_url = newUrl
        prod.action = 'fixed'
        console.log('ARREGLADO ✓')
      } catch (e) {
        errors++
        prod.action = 'db_error'
        console.log('DB ERROR')
      }
    } else {
      try {
        await pool.query('UPDATE products SET image_url = NULL WHERE id = $1', [prod.id])
        nulled++
        prod.action = 'nulled'
        console.log('SIN REEMPLAZO → null')
      } catch (e) {
        errors++
        prod.action = 'db_error'
        console.log('DB ERROR')
      }
    }

    done.add(prod.id)
    if ((i + 1) % 10 === 0) {
      saveProgress(done)
      saveBrokenLog(brokenList)
    }

    await sleep(1500 + Math.random() * 2000)

    if ((i + 1) % 50 === 0) {
      console.log(`\n  --- Progreso: ${fixed} arreglados, ${nulled} nulled, ${errors} errores ---\n`)
    }
  }

  saveProgress(done)
  saveBrokenLog(brokenList)

  const totalTime = ((Date.now() - start) / 1000 / 60).toFixed(1)
  console.log('\n' + '═'.repeat(60))
  console.log('  RESULTADO FINAL')
  console.log('═'.repeat(60))
  console.log(`  Tiempo:                ${totalTime} minutos`)
  console.log(`  Arregladas:            ${fixed}`)
  console.log(`  Sin reemplazo (null):  ${nulled}`)
  console.log(`  Errores DB:            ${errors}`)
  if (nulled > 0) console.log(`\n  → Ejecuta update_images_cloudsql.mjs para los ${nulled} productos sin imagen`)
  console.log('═'.repeat(60))

  await pool.end()
}

main().catch(err => {
  console.error('Error fatal:', err)
  process.exit(1)
})

File: pharmacy-ecommerce/scripts/fix_broken_images.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Detecta y arregla imágenes rotas en productos.
1. Obtiene todos los productos con image_url (no null/vacío)
2. Hace HEAD request a cada URL (timeout 6s)
3. Si la URL está caída (4xx, 5xx, timeout, error de conexión):
   - Busca imagen de reemplazo en DuckDuckGo
   - Actualiza Supabase con la nueva URL
   - Si no encuentra reemplazo, pone image_url = null para que otro script lo intente
4. Guarda progreso en broken_check_progress.json
"""
import sys
import io
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')

import time
import random
import json
import urllib.request
import urllib.parse
import urllib.error

try:
    from duckduckgo_search import DDGS
except ImportError:
    print("Instalando duckduckgo-search...")
    import subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "duckduckgo-search"])
    from duckduckgo_search import DDGS

SUPABASE_URL = 'https://jvagvjwrjiekaafpjbit.supabase.co'
SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imp2YWd2andyamlla2FhZnBqYml0Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3MDU5MDI5NCwiZXhwIjoyMDg2MTY2Mjk0fQ.wx7EcK6R5AT_ATd3wuXMoKW2NQJKyUimJmGrUrLfsAo'

PROGRESS_FILE = 'broken_check_progress.json'
BROKEN_LOG_FILE = 'broken_images_log.json'

# ─── Supabase helpers ────────────────────────────────────────────────────────

def supabase_get(table, query=''):
    url = f'{SUPABASE_URL}/rest/v1/{table}?{query}'
    req = urllib.request.Request(url, headers={
        'apikey': SUPABASE_KEY,
        'Authorization': f'Bearer {SUPABASE_KEY}',
    })
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.loads(resp.read())

def supabase_patch(table, query, data):
    url = f'{SUPABASE_URL}/rest/v1/{table}?{query}'
    req = urllib.request.Request(url, method='PATCH', headers={
        'apikey': SUPABASE_KEY,
        'Authorization': f'Bearer {SUPABASE_KEY}',
        'Content-Type': 'application/json',
        'Prefer': 'return=minimal',
    }, data=json.dumps(data).encode())
    with urllib.request.urlopen(req, timeout=15) as resp:
        return resp.status

# ─── Progress helpers ─────────────────────────────────────────────────────────

def load_progress():
    try:
        with open(PROGRESS_FILE, 'r') as f:
            return set(json.load(f))
    except Exception:
        return set()

def save_progress(done_ids):
    with open(PROGRESS_FILE, 'w') as f:
        json.dump(list(done_ids), f)

def load_broken_log():
    try:
        with open(BROKEN_LOG_FILE, 'r') as f:
            return json.load(f)
    except Exception:
        return []

def save_broken_log(entries):
    with open(BROKEN_LOG_FILE, 'w') as f:
        json.dump(entries, f, indent=2, ensure_ascii=False)

# ─── URL check ───────────────────────────────────────────────────────────────

def is_url_alive(url, timeout=6):
    """Returns (alive: bool, status_code: int|None)"""
    try:
        # Ensure https
        if url.startswith('http://'):
            url = 'https://' + url[7:]
        req = urllib.request.Request(
            url,
            method='HEAD',
            headers={
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
                'Accept': 'image/*,*/*',
            }
        )
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            status = resp.status
            return status < 400, status
    except urllib.error.HTTPError as e:
        return False, e.code
    except Exception:
        # Connection error, timeout, SSL, etc.
        return False, None

# ─── Image search ─────────────────────────────────────────────────────────────

def search_image(name, lab='', retry_count=0):
    clean_name = name.split('(')[0].strip()
    queries = [
        f"{clean_name} {lab} farmacia".strip(),
        f"{clean_name} medicamento",
        f"{clean_name} comprimido pastilla",
        clean_name[:40],
    ]
    seen = set()
    unique_queries = []
    for q in queries:
        if q.strip() and q not in seen:
            seen.add(q)
            unique_queries.append(q)

    for query in unique_queries:
        try:
            with DDGS() as ddgs:
                results = list(ddgs.images(query, region="cl-es", safesearch="off", max_results=5))
                for r in results:
                    img = r.get('image', '')
                    if (img.startswith('http') and 'x-raw-image' not in img
                            and len(img) < 500 and not img.endswith('.gif')):
                        return img
        except Exception as e:
            err_str = str(e).lower()
            if 'ratelimit' in err_str or '202' in str(e) or 'too many' in err_str:
                if retry_count < 2:
                    print('\n[RATE LIMIT] Esperando 30s...', flush=True)
                    time.sleep(30)
                    return search_image(name, lab, retry_count + 1)
            time.sleep(2)

    return None

# ─── Main ─────────────────────────────────────────────────────────────────────

def main():
    print('=' * 60)
    print('  DETECCIÓN Y ARREGLO DE IMÁGENES ROTAS')
    print('=' * 60)
    print()

    done_ids = load_progress()
    broken_log = load_broken_log()
    if done_ids:
        print(f'Retomando — {len(done_ids)} productos ya chequeados')

    # Fetch all products WITH an image_url (not null, not empty)
    print('Cargando productos con image_url...')
    products = []
    offset = 0
    while True:
        # Supabase filter: image_url not null AND not empty string
        batch = supabase_get(
            'products',
            f'select=id,name,laboratory,image_url'
            f'&image_url=not.is.null'
            f'&image_url=neq.'
            f'&order=name&limit=500&offset={offset}'
        )
        if not batch:
            break
        products.extend(batch)
        offset += 500
        if len(batch) < 500:
            break

    # Filter already checked
    products = [p for p in products if p['id'] not in done_ids]
    total = len(products)
    print(f'Total con imagen: {len(products) + len(done_ids)} | Por chequear: {total}\n')

    if total == 0:
        print('Todos los productos ya fueron chequeados.')
        return

    broken_count = 0
    fixed_count = 0
    nulled_count = 0
    ok_count = 0
    start = time.time()

    for i, prod in enumerate(products):
        pid = prod['id']
        name = prod['name']
        lab = prod.get('laboratory', '') or ''
        image_url = prod.get('image_url', '') or ''

        clean_name = name.split('(')[0].strip()[:38].ljust(38)
        elapsed = (time.time() - start) / 60
        eta = (elapsed / (i + 1)) * (total - i - 1) if i > 0 else 0
        print(f'[{i+1}/{total}] ({elapsed:.1f}m ETA:{eta:.1f}m) {clean_name} ', end='', flush=True)

        alive, status = is_url_alive(image_url)

        if alive:
            ok_count += 1
            print(f'OK ({status})')
        else:
            broken_count += 1
            status_str = str(status) if status else 'ERR'
            print(f'ROTO ({status_str}) → buscando...', end=' ', flush=True)

            # Log the broken URL
            broken_log.append({
                'id': pid,
                'name': name,
                'broken_url': image_url,
                'status': status,
            })

            # Search for replacement
            new_url = search_image(name, lab)
            if new_url:
                if new_url.startswith('http://'):
                    new_url = 'https://' + new_url[7:]
                try:
                    supabase_patch('products', f'id=eq.{pid}', {'image_url': new_url})
                    fixed_count += 1
                    print('ARREGLADO ✓')
                    broken_log[-1]['new_url'] = new_url
                    broken_log[-1]['action'] = 'fixed'
                except Exception:
                    print('DB ERROR')
                    broken_log[-1]['action'] = 'db_error'
            else:
                # No replacement found — null out so update_images_supabase.py can retry
                try:
                    supabase_patch('products', f'id=eq.{pid}', {'image_url': None})
                    nulled_count += 1
                    print('SIN REEMPLAZO → null')
                    broken_log[-1]['action'] = 'nulled'
                except Exception:
                    print('DB ERROR')
                    broken_log[-1]['action'] = 'db_error'

            time.sleep(random.uniform(1.5, 3.0))

        # Save progress every 10 items
        done_ids.add(pid)
        if (i + 1) % 10 == 0:
            save_progress(done_ids)
            save_broken_log(broken_log)

        # Summary every 100
        if (i + 1) % 100 == 0:
            print(f'\n--- OK: {ok_count} | Rotas: {broken_count} (arregladas: {fixed_count}, nulled: {nulled_count}) ---\n')

    save_progress(done_ids)
    save_broken_log(broken_log)
    total_time = (time.time() - start) / 60

    print()
    print('=' * 60)
    print('  RESULTADO FINAL')
    print('=' * 60)
    print(f'  Tiempo:              {total_time:.1f} minutos')
    print(f'  URLs válidas:        {ok_count}')
    print(f'  URLs rotas:          {broken_count}')
    print(f'  Arregladas:          {fixed_count}')
    print(f'  Sin reemplazo (null): {nulled_count}')
    print(f'  Log guardado en:     {BROKEN_LOG_FILE}')
    print('=' * 60)
    print()
    if nulled_count > 0:
        print(f'  → Ejecuta update_images_supabase.py para los {nulled_count} productos sin reemplazo')

if __name__ == '__main__':
    main()
